tests/src/providers/testqgsvirtualrasterprovider.cpp

```diff
@@ -0,0 +1,232 @@
+  QFileInfo landsatRasterFileInfo( landsatFileName );
+  mLandsatRasterLayer = new QgsRasterLayer( landsatRasterFileInfo.filePath(),
+      landsatRasterFileInfo.completeBaseName() );
+}
+
+void TestQgsVirtualRasterProvider::validLayer()
+{
+  QgsRasterLayer::LayerOptions options;
+
+  std::unique_ptr< QgsRasterLayer > layer = std::make_unique< QgsRasterLayer >(
+        mTestDataDir + QStringLiteral( "raster/dem.tif" ),
+        QStringLiteral( "layer" ),
+        QStringLiteral( "virtualraster" ),
+        options
+      );
+
+  QVERIFY( ! layer->isValid() );
+}
+
+//runs after all tests
+void TestQgsVirtualRasterProvider::cleanupTestCase()
+{
+  QgsApplication::exitQgis();
+  QString myReportFile = QDir::tempPath() + "/qgistest.html";
+  QFile myFile( myReportFile );
+  if ( myFile.open( QIODevice::WriteOnly | QIODevice::Append ) )
+  {
+    QTextStream myQTextStream( &myFile );
+    myQTextStream << mReport;
+    myFile.close();
+  }
+}
+
+void TestQgsVirtualRasterProvider::testUriProviderDecoding()
+{
+
+  QgsRasterDataProvider::VirtualRasterParameters decodedParams = QgsVirtualRasterProvider::decodeVirtualRasterProviderUri( QStringLiteral( "?crs=EPSG:4326&extent=18.6662979442000001,45.7767014376000034,18.7035979441999984,45.8117014376000000&width=373&height=350&formula=\"dem@1\" + 200&dem:uri=path/to/file&dem:provider=gdal&landsat:uri=path/to/landsat&landsat:provider=gdal" ) );
+
+  QCOMPARE( decodedParams.width, 373 );
+  QCOMPARE( decodedParams.height, 350 );
+  QCOMPARE( decodedParams.extent, QgsRectangle( 18.6662979442000001, 45.7767014376000034, 18.7035979441999984, 45.8117014376000000 ) );
+  QCOMPARE( decodedParams.crs, QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ) );
+  QCOMPARE( decodedParams.formula, QStringLiteral( "\"dem@1\" + 200" ) );
+
+  QCOMPARE( decodedParams.rInputLayers.at( 1 ).provider, QStringLiteral( "gdal" ) );
+  QCOMPARE( decodedParams.rInputLayers.at( 0 ).provider, QStringLiteral( "gdal" ) );
+
+  qDebug() << QStringLiteral( "Raster layer: name, uri, provider" );
+  for ( int i = 0; i < decodedParams.rInputLayers.size() ; ++i )
+  {
+    qDebug() << decodedParams.rInputLayers.at( i ).name << ", " <<
+             decodedParams.rInputLayers.at( i ).uri  << ", " <<
+             decodedParams.rInputLayers.at( i ).provider;
+  }
+
+}
+
+void TestQgsVirtualRasterProvider::testUriEncoding()
+{
+
+  QgsRasterDataProvider::VirtualRasterParameters decodedParams = QgsVirtualRasterProvider::decodeVirtualRasterProviderUri( QStringLiteral( "?crs=EPSG:4326&extent=18.6662979442000001,45.7767014376000034,18.7035979441999984,45.8117014376000000&width=373&height=350&formula=\"dem@1\" + 200&dem:uri=/home/franc/dev/cpp/QGIS/tests/testdata/raster/dem.tif&dem:provider=gdal&rband:uri=/home/franc/dev/cpp/QGIS/tests/testdata/raster/band1_byte_ct_epsg4326.tif&rband:provider=gdal" ) );
+  qDebug() << QgsVirtualRasterProvider::encodeVirtualRasterProviderUri( decodedParams );
```

Comment:
let's modify this test to have automatic testing: create `QgsRasterDataProvider::VirtualRasterParameters` struct and manually populate its values (`params.width = 600;` and so on), then call encodeVirtualRasterProviderUri() and check the returned URI against expected string.